### PR TITLE
Introduced plugin for generator-level tt+X event categorisation

### PR DESCRIPTION
**Pull request by bartosik-hep**
Allows to separate between different tt+jets subprocesss with a single integer number: tt+bb, tt+cc, tt+2b, tt+2c, tt+light jet, etc.
An improved version of the custom producer provided in the [TWiki](https://twiki.cern.ch/twiki/bin/view/CMSPublic/GenHFHadronMatcher#Event_categorization_example_1), based on the [GenHFHadronMatcher](https://twiki.cern.ch/twiki/bin/view/CMSPublic/GenHFHadronMatcher) tool.
Supports all decay channels of the ttbar system.
To be used by Top and ttH analyses.

### CODE REVIEW

**Comment by monttj** [comment]
please test

**Comment by kreczko** [comment]
+1

**Comment by monttj** [comment]
+1

File: TopQuarkAnalysis/TopTools/plugins/GenTtbarCategorizer.cc [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+void
+GenTtbarCategorizer::endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&)
+{
+}
+*/
+ 
+// ------------ method returns a vector of jet indices from the given map, sorted by N hadrons in descending order  ------------
+std::vector<int> GenTtbarCategorizer::nHadronsOrderedJetIndices(const std::map<int, int>& m_jetIndex)
+{
+    std::vector<std::pair<int, int> > v_jetNhadIndexPair;
```
Comment (davidlange6) [inline TopQuarkAnalysis/TopTools/plugins/GenTtbarCategorizer.cc]: suggest to add a v_jetNhadIndexPair.reserve(m_jetIndex.size()); here

File: TopQuarkAnalysis/TopTools/plugins/GenTtbarCategorizer.cc [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+        const int jetIndex = it->first;
+        const int nHadrons = it->second;
+        v_jetNhadIndexPair.push_back( std::pair<int, int>(nHadrons, jetIndex) );
+    }
+    // Sorting the vector of pairs by their key value
+    std::sort(v_jetNhadIndexPair.begin(), v_jetNhadIndexPair.end(), std::greater<std::pair<int, int> >());
+    // Building the vector of indices in the proper order
+    std::vector<int> orderedJetIndices;
+    for(std::vector<std::pair<int, int> >::const_iterator it = v_jetNhadIndexPair.begin(); it != v_jetNhadIndexPair.end(); ++it) {
+        orderedJetIndices.push_back(it->second);
```
Comment (davidlange6) [inline TopQuarkAnalysis/TopTools/plugins/GenTtbarCategorizer.cc]: same sort of comment of orderedJetIndices

**Comment by davidlange6** [comment]
@cmsbuild, please test

**Comment by monttj** [comment]
+1 
try again

**Comment by davidlange6** [comment]
@cmsbuild, please test 
(there are commits here)

**Comment by monttj** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**